src/main/webapp/app/course/learning-paths/participate/learning-path-storage.service.ts

```diff
@@ -0,0 +1,211 @@
+        }
+        return this.learningPathHistories.get(learningPathId)!.pop();
+    }
+
+    /**
+     * Simplifies and stores the recommended order of learning objects for the given learning path
+     *
+     * @param learningPathId the id of the learning path
+     * @param learningPath the learning path dto that should be stored
+     */
+    storeRecommendations(learningPathId: number, learningPath: NgxLearningPathDTO) {
+        this.learningPathRecommendations.set(learningPathId, []);
+        let currentId = learningPath.nodes.map((node) => node.id).find((id) => !learningPath.edges.find((edge) => edge.target == id));
+        while (currentId) {
+            const currentNode = learningPath.nodes.find((node) => node.id == currentId)!;
+            if (currentNode.type === NodeType.LECTURE_UNIT) {
+                this.learningPathRecommendations.get(learningPathId)!.push(new LectureUnitEntry(currentNode.linkedResourceParent!, currentNode.linkedResource!));
+            } else if (currentNode.type === NodeType.EXERCISE) {
+                this.learningPathRecommendations.get(learningPathId)!.push(new ExerciseEntry(currentNode.linkedResource!));
+            }
+            const edge = learningPath.edges.find((edge) => edge.source == currentId);
+            if (edge) {
+                currentId = edge.target;
+            } else {
+                currentId = undefined;
+            }
+        }
+    }
+
+    /**
+     * Returns if there is a recommendation left that has not been interacted with yet.
+     *
+     * @param learningPathId the id of the learning path
+     */
+    hasRecommendation(learningPathId: number) {
+        if (this.learningPathRecommendations.has(learningPathId)) {
+            return this.learningPathRecommendations.get(learningPathId)!.find((entry) => !entry.interacted) !== undefined;
+        }
+        return false;
+    }
+
+    /**
+     * Sets the given entry interaction.
+     *
+     * @param learningPathId the id of the learning path the entry belongs to
+     * @param entry the entry that should be set to interacted with
+     * @param value the value that should be set
+     */
+    setInteraction(learningPathId: number, entry: StorageEntry, value: boolean) {
+        if (!this.learningPathRecommendations.has(learningPathId)) {
+            return;
+        }
+        const storedEntry = this.getStoredRecommendationEntry(learningPathId, entry);
+        if (storedEntry) {
+            storedEntry.interacted = value;
+        }
+    }
+
+    /**
+     * Gets the next recommended entry for a learning object.
+     * <p>
+     * If the given entry has no successor, the first entry that has not been interacted with will be returned.
+     * @param learningPathId the id of the learning path
+     * @param entry the entry for which the successor should be returned
+     */
+    getNextRecommendation(learningPathId: number, entry: StorageEntry | undefined): StorageEntry | undefined {
```

Comment:
```suggestion
    getNextRecommendation(learningPathId: number, entry?: StorageEntry): StorageEntry | undefined {
```
I'm not sure about that change. Do you want to force the second parameter, or can it be optional, since its absence is handled in this method